### PR TITLE
Migrating to SARIF Schema

### PR DESCRIPTION
**Pull request by carakessler**
Migrating all `Tasks` to return `Result[]` (from [sarif](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/289ba098a611b80ee03ed72a2feee41a593fe765/types/sarif/index.d.ts#L247)) and migrating the run command to return a sarif `Log`

For more details, check out the RFC https://github.com/checkupjs/rfcs/pull/2

### CODE REVIEW

**Comment by carakessler** [comment]
I ran checkup JSON results through a JSON schema validator to ensure it is complying with SARIF and ran into a problem. The field `occurrenceCount` [is meant be an integer with value >= 1](https://raw.githubusercontent.com/schemastore/schemastore/master/src/schemas/json/sarif-2.1.0-rtm.5.json#/definitions/result/properties/occurrenceCount/minimum), meaning there can not  be a result with `occurrenceCount = 0`. We were using an `occurrenceCount = 0` to model Results from Tasks where we didn't find what we were looking for - most straightforward example is `eslint-disables` and `template-lint-disables`; in those tasks we are looking for instances of people overriding the linter, and in the case that we don't find any, how should we communicate that? I came up with the following options: 

1. Just deal with the loss of explicit data. Infer that when there is no Result returned from a Task it means we didnt find what we were looking for, and so therefore there were zero of them. 

	Pros:
	- complies to SARIF constructs	
	
	Cons: 

	- adds technical complexity because it requires us to break the assumption that a Task will always return a Result
	- makes the shape of the JSON result schema nondeterministic (sometimes we will have Results for task _foo_, sometimes we wont
	- no way of knowing what Tasks were run (if there were tasks that produced 0 results) 


2. Basically option 1 but also add a new construct to checkup where we store "things we were looking for and didn't find" and then throw it into the propertyBag for the invocation. So when a task goes and looks for something and finds 0 instances, it doesnt return a Result, rather it simply pushes the name of the task (or maybe a more descriptive string) to some global object and then we can do whatever we want with those at the end. 
	Pros:
	- complies to SARIF constructs

	Cons: 
	- adds a new (kind of confusing) construct 
	- adds technical complexity because it requires us to break the assumption that a Task will always return a Result 
	- makes the shape of the JSON result schema nondeterministic (sometimes we will have Results for task _foo_, sometimes we wont BUT at least we will have a way to derive what Tasks returned 0 results from any JSON output.


3. `occurrenceCount` is an optional field on a `Result`, so in the cases we don't find what we are looking for, instead of just putting `occurrenceCount = 0`, just omit the field entirely. We will infer internally that when `occurrenceCount` is not populated, it means there are no instances of what we were looking for 
	Pros: 
	- complies to SARIF constructs
	- requires minimal refactoring

	Cons: 
	- since `occurrenceCount` is an optional field, its (in theory) shouldn't lead a consumer of this data to assume that it's 0, it could be omitted for other reasons. This "implicit" zero could lead to other bugs/problems and is a bit confusing. 

4. Leave everything as is. This is all an edge case, and the SARIF spec will be adhered to the vast majority of the time, and although this technically breaks the spec, it is doing so in a pretty sane and straightforward way. We aren't putting a string where there should be a number or totally misusing fields, we are just using a number that the spec doesn't want in an edge case. This is unlikely to break downstream consumers of SARIF data 
	Pros: 
	- most straightforward in terms of implementation 
	- produces the most explicit and descriptive results model
	
	Cons: 
	- not 100% adherent with the SARIF spec 100% of the time.
	- the lack of adherence could be confusing for those familiar with SARIF, but not our implementation

I vote for option 3.

**Review by scalvert — APPROVED** [review]

**Comment by scalvert** [comment]
Thanks for working on this huge refactor! Looks great!!

File: packages/cli/src/reporters/console-reporter.ts [hunk 1]
```diff
@@ -1,136 +1,152 @@
...
+  getRegisteredTaskReporters,
   ui,
-  Result,
+  groupDataByField,
+  NO_RESULTS_FOUND,
 } from '@checkup/core';
-import { yellow, bold } from 'chalk';
-import TaskList from '../task-list';
+import * as chalk from 'chalk';
+import { bold, yellow } from 'chalk';
```
[the remark this answers is not in the document]
Comment (carakessler) [reply packages/cli/src/reporters/console-reporter.ts:11]: fixed

File: packages/checkup-plugin-javascript/src/actions/eslint-summary-actions.ts [hunk 2]
```diff
@@ -1,17 +1,14 @@
...
-  let warnings = taskResult.result.find(
-    (result: MultiValueResult) => result.key === 'eslint-warnings'
-  )!;
+  let errors = taskResults.filter((result: Result) => result.properties?.type === 'error')!;
+  let warnings = taskResults.filter((result: Result) => result.properties?.type === 'warning')!;
 
-  let errorCount = errors.dataSummary.total;
-  let warningCount = warnings.dataSummary.total;
+  let errorCount = errors.reduce((total, error) => total + (error.occurrenceCount || 0), 0);
+  let warningCount = warnings.reduce((total, warning) => total + (warning.occurrenceCount || 0), 0);
```
[the remark this answers is not in the document]
Comment (carakessler) [reply packages/checkup-plugin-javascript/src/actions/eslint-summary-actions.ts:11]: SCREAMING :) 

ya maybe to be more specific a `sumOccurrences`, as `occurrenceCount` is an optional field. ill throw that in

**Comment by carakessler** [comment]
Added `sumOccurrences` function (as per @scalvert's recommendation) and ensured that json output passes sarif validation when running on travis-web, going to merge

**Pull request merged**